internal/handler/report_api_csp.go

```diff
@@ -0,0 +1,183 @@
+		if gotMetadata {
+			metadata = metadatas[0]
+		}
+	}
+
+	for _, violation := range reports.Reports {
+		report_only := violation.Body.Disposition == "report"
+		lf := log.Fields{
+			"report_only":         report_only,
+			"document_uri":        violation.Body.DocumentURL,
+			"referrer":            violation.Body.Referrer,
+			"blocked_uri":         violation.Body.BlockedURL,
+			"violated_directive":  violation.Body.EffectiveDirective,
+			"effective_directive": violation.Body.EffectiveDirective,
+			"original_policy":     violation.Body.OriginalPolicy,
+			"disposition":         violation.Body.Disposition,
+			"status_code":         violation.Body.StatusCode,
+			"source_file":         violation.Body.SourceFile,
+			"line_number":         violation.Body.LineNumber,
+			"column_number":       violation.Body.ColumnNumber,
+			"metadata":            metadata,
+			"path":                r.URL.Path,
+		}
+
+		if vrh.TruncateQueryStringFragment {
+			lf["document_uri"] = utils.TruncateQueryStringFragment(violation.Body.DocumentURL)
+			lf["referrer"] = utils.TruncateQueryStringFragment(violation.Body.Referrer)
+			lf["blocked_uri"] = utils.TruncateQueryStringFragment(violation.Body.BlockedURL)
+			lf["source_file"] = utils.TruncateQueryStringFragment(violation.Body.SourceFile)
+		}
+
+		if vrh.LogClientIP {
+			ip, err := utils.GetClientIP(r)
+			if err != nil {
+				vrh.Logger.Warnf("unable to parse client ip: %s", err)
+			}
+			lf["client_ip"] = ip.String()
+		}
+
+		if vrh.LogTruncatedClientIP {
+			ip, err := utils.GetClientIP(r)
+			if err != nil {
+				vrh.Logger.Warnf("unable to parse client ip: %s", err)
+			}
+			lf["client_ip"] = utils.TruncateClientIP(ip)
+		}
+
+		vrh.Logger.WithFields(lf).Info()
+	}
+
+}
+
+func (vrh *ReportAPIViolationReportHandler) validateViolation(r ReportAPIReports) error {
+	for _, violation := range r.Reports {
+		if violation.Type != "csp-violation" {
+			continue // Skip the rest of the loop and move to the next iteration
+		}
+		for _, value := range vrh.BlockedURIs {
+			if strings.HasPrefix(violation.Body.BlockedURL, value) {
+				err := fmt.Errorf("blocked URI ('%s') is an invalid resource", value)
+				return err
+			}
+		}
+		if !strings.HasPrefix(violation.Body.DocumentURL, "http") {
+			return fmt.Errorf("document URI ('%s') is invalid", violation.Body.DocumentURL)
+		}
+	}
+
+	return nil
+}
+
+func ReportAPICorsHandler(w http.ResponseWriter, r *http.Request) {
+	origin := r.Header.Get("Origin")
+	method := r.Header.Get("Access-Control-Request-Method")
+	header := r.Header.Get("Access-Control-Request-Headers")
+	allow_origin := utils.Ternary(origin != "", origin, "*")
+	allow_method := utils.Ternary(method != "", method, "*")
+	allow_header := utils.Ternary(header != "", header, "*")
+	// Special handling due to bug in Chrome
+	// https://bugs.chromium.org/p/chromium/issues/detail?id=1152867
+	w.Header().Set("Access-Control-Allow-Origin", allow_origin)
```

Comment:
we shouldn't blindly accept the value here; you'll want to do some validation around the value we receive and return in the header.

Reply:
@jacobbednarz what kind of validation did you have in mind? I mean it is just for CORS. Most other CSP report collection services like report-uri.com use "*" every time. 
I can add a validation but what should be returned if it fails?

Reply:
i would go with is a valid domain or `*` - namely just looking for something to avoid being chained in a SSRF attack if we ever end up using this value anywhere.